### PR TITLE
Allow user to specify DataSet event severity field

### PR DESCRIPTION
**Pull request by Kami**
## Description

This pull request allows users to set  top level``severity`` (sev) field on the DataSet event.

This field contains the event severity / log level and it's special since it's a top level event field (same as ``id``, ``thread`` and ``log`` event attribute) and not part of the event attributes aka ``event.attrs`` field.

## Usage

To use this functionality, user needs to configure ``severity_field`` plugin config option. This config option tells the plugin which Logstash event field contains the severity value.

For example:

```
output {
 scalyr {
   api_write_token => 'SCALYR_API_KEY'
   ...
   severity_field => 'severity',
 }
}
```

In this example, ``severity`` Logstash event field value carries the severity value. Actual field value must be an integer between 0 to 6 (inclusive), from finest / most verbose to the most critical / least verbose.

This numeric severity field value is mapped to a severity level on the DataSet server side as shown below:

- 0 -> finest
- 1 -> trace
- 2 -> debut
- 3 -> info
- 4 -> warning
- 5 -> error
- 6 -> fatal / emergency / critical

## Notes

For backward compatibility reasons this functionality is disabled by default (``severity_field`` plugin config option value defaults to ``nil``). It needs to be explicitly enabled by the end user (opt-int) by configuring ``severity_field`` plugin config option.

### CODE REVIEW

**Comment by Kami** [comment]
NOTE: stable end to end tests will continue to fail until this PR is merged and new version of the plugin published to RubyGems.

File: spec/logstash/outputs/scalyr_spec.rb [hunk 1]
```diff
@@ -225,6 +255,86 @@ def post_add_events(body, is_status, body_serialization_duration = 0)
...
+        (0..2).each do |index|
+          expect(body['events'][index]['attrs'].fetch('severity', nil)).to eq(nil)
+          expect(body['events'][index]['attrs'].fetch('sev', nil)).to eq(nil)
+          expect(body['events'][index]['sev']).to eq(nil)
+        end
+      end
+
+      it "works correctly when severity event attribute is not specified but severity field is not set" do
+        # Since severity_field config option is not set, severity field should be treated as a
+        # regulat event attribute and not as s a special top level event field
```
Comment (ArthurKamalov) [inline spec/logstash/outputs/scalyr_spec.rb:319]: typo

File: lib/logstash/outputs/scalyr.rb [hunk 2]
```diff
@@ -60,6 +60,22 @@ class LogStash::Outputs::Scalyr < LogStash::Outputs::Base
...
+  #
+  # - 0 -> finest
+  # - 1 -> trace
+  # - 2 -> debut
+  # - 3 -> info
+  # - 4 -> warning
+  # - 5 -> error
+  # - 6 -> fatal / emergency / critical
+  #
+  # By default, if Event contains no severity field, default value of 3 (info) will be used.
```
Comment (ArthurKamalov) [inline lib/logstash/outputs/scalyr.rb:76]: Just to clarify for me: 
> default value of 3 (info) will be used.

Event - DataSet event, the default value is set on server side?

**Review by ArthurKamalov — APPROVED** [review]
It took some time for me to fully remember how things work here :), but now everything looks very straightforward to me 👍

File: lib/logstash/outputs/scalyr.rb (hunk 2 above)
Comment (Kami) [reply lib/logstash/outputs/scalyr.rb:76]: Correct, if not provided as part of the Event.sev field, server-side sets the value to 3 (info).

File: spec/logstash/outputs/scalyr_spec.rb (hunk 1 above)
Comment (Kami) [reply spec/logstash/outputs/scalyr_spec.rb]: Fixed, thanks.

File: lib/logstash/outputs/scalyr.rb [hunk 3]
```diff
@@ -868,9 +877,9 @@ def build_multi_event_request_array(logstash_events)
         scalyr_event[:log] = logs_ids[log_identifier]
       end
 
-      # optionally set severity (if available)
+      # optionally set severity (if available and valid)
       if @severity_field and severity
```
[the remark this answers is not in the document]
Comment (Kami) [reply lib/logstash/outputs/scalyr.rb:881]: That's a good catch, thanks.

I thought I have added test cases for all the edge case, but it looks like that may not be the case. Will make sure I also add a test case for this scenario.

[the remark this answers is not in the document]
Comment (Kami) [reply lib/logstash/outputs/scalyr.rb]: Fixed in - 8dfe68b.

Existing tests were not strict enough - they didn't differentiate between value not being set and value being set to `nil` (another +1 for strict types again :P)

**Review by ArthurKamalov — APPROVED** [review]
